diff --git a/tests/helpers.py b/tests/helpers.py
index c96f512..e9f1da3 100644
--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -15,6 +15,7 @@

 # pylint: disable=missing-class-docstring,missing-function-docstring,invalid-name

+import contextlib
 import dataclasses
 import gc
 import itertools
@@ -67,6 +68,18 @@ def parametrize(**argvalues):
     return pytest.mark.parametrize(arguments, argvalues, ids=ids)


+@contextlib.contextmanager
+def recursionlimit(limit):
+    gc_collect()
+    old_limit = sys.getrecursionlimit()
+    sys.setrecursionlimit(min(old_limit, limit))
+    try:
+        yield
+    finally:
+        sys.setrecursionlimit(old_limit)
+        gc_collect()
+
+
 MISSING = object()


diff --git a/tests/test_concurrent.py b/tests/test_concurrent.py
index d15c928..dac0c2f 100644
--- a/tests/test_concurrent.py
+++ b/tests/test_concurrent.py
@@ -25,7 +25,15 @@ from concurrent.futures import ThreadPoolExecutor, as_completed
 import pytest

 import optree
-from helpers import GLOBAL_NAMESPACE, PYPY, TREES, Py_GIL_DISABLED, gc_collect, parametrize
+from helpers import (
+    GLOBAL_NAMESPACE,
+    PYPY,
+    TREES,
+    Py_GIL_DISABLED,
+    gc_collect,
+    parametrize,
+    recursionlimit,
+)


 if PYPY:
@@ -246,8 +254,9 @@ def test_treespec_self_referential():  # noqa: C901

     concurrent_run(check7)

-    with pytest.raises(RecursionError):
-        assert treespec != other
+    with recursionlimit(64):
+        with pytest.raises(RecursionError):
+            assert treespec != other

     wr = weakref.ref(treespec)
     del treespec, key, other
diff --git a/tests/test_treespec.py b/tests/test_treespec.py
index 41152da..dcdefd2 100644
--- a/tests/test_treespec.py
+++ b/tests/test_treespec.py
@@ -43,6 +43,7 @@ from helpers import (
     MyDict,
     gc_collect,
     parametrize,
+    recursionlimit,
 )


@@ -246,8 +247,9 @@ def test_treespec_self_referential():
     assert hash(treespec) == hash(other)

     if not PYPY:
-        with pytest.raises(RecursionError):
-            assert treespec != other
+        with recursionlimit(64):
+            with pytest.raises(RecursionError):
+                assert treespec != other

     wr = weakref.ref(treespec)
     del treespec, key, other
